File: backend/app/routers/campaigns.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.dependencies.auth import get_current_user
from app.schemas.campaign import CampaignCreate, CampaignRead, CampaignStats, CampaignUpdate
from app.services import campaign_service
from app.tasks.send_campaign import run_send_campaign

router = APIRouter(prefix="/campaigns", tags=["campaigns"], dependencies=[Depends(get_current_user)])


@router.get("", response_model=list[CampaignRead])
async def list_campaigns(db: AsyncSession = Depends(get_db)):
    return await campaign_service.get_campaigns(db)


@router.post("", response_model=CampaignRead, status_code=201)
async def create_campaign(data: CampaignCreate, db: AsyncSession = Depends(get_db)):
    return await campaign_service.create_campaign(db, data)


@router.get("/{campaign_id}", response_model=CampaignRead)
async def get_campaign(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        return await campaign_service.get_campaign_read(db, campaign_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.put("/{campaign_id}", response_model=CampaignRead)
async def update_campaign(campaign_id: int, data: CampaignUpdate, db: AsyncSession = Depends(get_db)):
    try:
        return await campaign_service.update_campaign(db, campaign_id, data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{campaign_id}", status_code=204)
async def delete_campaign(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        await campaign_service.delete_campaign(db, campaign_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/{campaign_id}/start", response_model=CampaignRead)
async def start_campaign(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        campaign = await campaign_service.start_campaign(db, campaign_id)
        run_send_campaign.delay(campaign_id)
        return campaign
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{campaign_id}/pause", response_model=CampaignRead)
async def pause_campaign(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        return await campaign_service.pause_campaign(db, campaign_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{campaign_id}/stop", response_model=CampaignRead)
async def stop_campaign(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        return await campaign_service.stop_campaign(db, campaign_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{campaign_id}/stats", response_model=CampaignStats)
async def campaign_stats(campaign_id: int, db: AsyncSession = Depends(get_db)):
    try:
        return await campaign_service.get_campaign_stats(db, campaign_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/api/campaigns.ts
import api from "./client";
import type { Campaign, CampaignCreate, CampaignStats } from "@/types/campaign";

export const getCampaigns = () =>
  api.get<Campaign[]>("/campaigns").then((r) => r.data);

export const getCampaign = (id: number) =>
  api.get<Campaign>(`/campaigns/${id}`).then((r) => r.data);

export const createCampaign = (data: CampaignCreate) =>
  api.post<Campaign>("/campaigns", data).then((r) => r.data);

export const updateCampaign = (id: number, data: Partial<CampaignCreate>) =>
  api.put<Campaign>(`/campaigns/${id}`, data).then((r) => r.data);

export const deleteCampaign = (id: number) =>
  api.delete(`/campaigns/${id}`);

export const startCampaign = (id: number) =>
  api.post<Campaign>(`/campaigns/${id}/start`).then((r) => r.data);

export const pauseCampaign = (id: number) =>
  api.post<Campaign>(`/campaigns/${id}/pause`).then((r) => r.data);

export const stopCampaign = (id: number) =>
  api.post<Campaign>(`/campaigns/${id}/stop`).then((r) => r.data);

export const getCampaignStats = (id: number) =>
  api.get<CampaignStats>(`/campaigns/${id}/stats`).then((r) => r.data);
